### PR TITLE
Append parent name for nested functions while mangling

### LINKED ISSUE #2347: duplicate symbol for statement function while linking with clang
```fortran
subroutine a()
integer :: x
logical qsmall
qsmall(x) = .NOT. (sum .GE. abstol .AND. x .GE. eps*sum)
end
````

```fortran
subroutine b()
integer :: x
logical qsmall
qsmall(x) = .NOT. (sum .GE. abstol .AND. x .GE. eps*sum)
end
```

```console
% lfortran -c b.f90 -o a2.o --generate-object-code --implicit-typing --no-warnings                          
% lfortran -c a.f90 -o a1.o --generate-object-code --implicit-typing --no-warnings                          
% clang -shared -fPIC -o lib.so a1.o a2.o -L/Users/pranavchiku/repos/lfortran/src/runtime -llfortran_runtime

duplicate symbol '_qsmall' in:
    a1.o
    a2.o
ld: 1 duplicate symbol for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #2347.

```fortran
subroutine a()
integer :: x
logical qsmall
qsmall(x) = .NOT. (sum .GE. abstol .AND. x .GE. eps*sum)
end
```

> % lfortran a.f90 --show-asr --pass=unique_symbols --implicit-typing --all-mangling --mangle-underscore

<details>

<summary> ASR </summary>

```clojure
(TranslationUnit
    (SymbolTable
        1
        {
            a_:
                (Function
                    (SymbolTable
                        2
                        {
                            qsmall_a_:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            abstol_a_:
                                                (Variable
                                                    3
                                                    abstol_a_
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            eps_a_:
                                                (Variable
                                                    3
                                                    eps_a_
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            qsmall_return_var_name_a_:
                                                (Variable
                                                    3
                                                    qsmall_return_var_name_a_
                                                    []
                                                    ReturnVar
                                                    ()
                                                    ()
                                                    Default
                                                    (Logical 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            sum_a_:
                                                (Variable
                                                    3
                                                    sum_a_
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Real 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            x_a_:
                                                (Variable
                                                    3
                                                    x_a_
                                                    []
                                                    In
                                                    ()
                                                    ()
                                                    Default
                                                    (Integer 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    qsmall_a_
                                    (FunctionType
                                        [(Integer 4)]
                                        (Logical 4)
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    [(Var 3 x_a_)]
                                    [(=
                                        (Var 3 qsmall_return_var_name_a_)
                                        (LogicalNot
                                            (LogicalBinOp
                                                (RealCompare
                                                    (Var 3 sum_a_)
                                                    GtE
                                                    (Var 3 abstol_a_)
                                                    (Logical 4)
                                                    ()
                                                )
                                                And
                                                (RealCompare
                                                    (Cast
                                                        (Var 3 x_a_)
                                                        IntegerToReal
                                                        (Real 4)
                                                        ()
                                                    )
                                                    GtE
                                                    (RealBinOp
                                                        (Var 3 eps_a_)
                                                        Mul
                                                        (Var 3 sum_a_)
                                                        (Real 4)
                                                        ()
                                                    )
                                                    (Logical 4)
                                                    ()
                                                )
                                                (Logical 4)
                                                ()
                                            )
                                            (Logical 4)
                                            ()
                                        )
                                        ()
                                    )]
                                    (Var 3 qsmall_return_var_name_a_)
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            x_:
                                (Variable
                                    2
                                    x_
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    a_
                    (FunctionType
                        []
                        ()
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        .false.
                    )
                    []
                    []
                    []
                    ()
                    Public
                    .false.
                    .false.
                    ()
                )
        })
    []
)
```

</details>

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Minimal code to reproduce scipy failure

```fortran
subroutine a()
integer :: x
logical qsmall
qsmall(x) = .NOT. (sum .GE. abstol .AND. x .GE. eps*sum)
call b()
end subroutine
```

```console
% lfortran a.f90 --show-asr --pass=unique_symbols --implicit-typing --all-mangling --mangle-underscore --implicit-interface --bindc-mangling
```

This happens because of implicit interface functions get treated as nested and were  mangled with parent function name.

**Comment by Pranavchiku** [comment]
I tested locally, everything works now.

File: src/libasr/pass/unique_symbols.cpp [hunk 1]
```diff
@@ -109,7 +132,13 @@ class SymbolRenameVisitor: public ASR::BaseWalkVisitor<SymbolRenameVisitor> {
             }
         }
         for (auto &a : x.m_symtab->get_scope()) {
+            if (is_nested_function(a.second)) {
```
Comment (Smit-create) [inline src/libasr/pass/unique_symbols.cpp:135]: Instead of calling the same function twice, store the result in a variable.

**Comment by certik** [comment]
@Smit-create does it look good otherwise?

**Comment by Smit-create** [comment]
Yeah, I think it's fine.

**Pull request merged**